Make remote sharee search configurable

michielbdejong opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.com/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next version of ownCloud.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" if the PR still has open tasks
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
When the 'sciencemesh' app is installed, use it instead of the federatedfilesharing app to
find sharee matches for OCM sharing.

## Related Issue
- Fixes #40576

## Motivation and Context
This is needed for OC-10 instances that want to join https://sciencemesh.io

## How Has This Been Tested?
- test environment:
  - GitPod from https://github.com/cs3org/ocm-test-suite
  - run ./gitpod-init.sh
  - run ./scripts/clean.sh
  - run ./scripts/orro-testing.sh
  - go the Firefox browser-in-a-browser that is exposed to port 5800
  - in there, navigate to https://oc1.docker (einstein / relativity)
  - create a local user 'john'
  - using the ScienceMesh app, generate an invite
  - accept the invite on oc2.docker (marie / radioactivity)
  - confirm that marie is now a ScienceMesh contact of einstein on oc1.docker
- test case 1: sharing via ScienceMesh
  - as einstein on oc1.docker, share a file and start typing "m..." as the sharee
  - as marie on oc2.docker, accept the share
  - check the revaoc1.docker logs and see that the sharing happened via oc1 -> revaoc1 -> revaoc2 -> oc2
- test case 2: sharing via local
  - as einstein on oc1.docker, share a file and start typing "jo..." as the sharee
  - it should offer to share with John as a local user

## Screenshots (if appropriate):
<img width="1656" alt="Screenshot 2023-01-13 at 13 53 37" src="https://user-images.githubusercontent.com/408412/212324749-155a6ba1-53de-43fc-8ad8-c3ba6fcf2c06.png">

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [x] Code changes
- [ ] ~~Unit tests added~~
- [ ] ~~Acceptance tests added~~
- [ ] ~~Documentation ticket raised~~
- [x] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

michielbdejong committed:
OCM via ScienceMesh

Signed-off-by: Michiel de Jong <michiel@unhosted.org>

michielbdejong commented:
I'll try to add some tests!

michielbdejong commented:
phpstan complains: "Instantiated class OCA\ScienceMesh\Plugins\ScienceMeshSearchPlugin not found."
It would be nice if phpstan would support a way to disable this warning at this specific line.
Not sure how to deal with this, will try to think of a solution.

michielbdejong commented:
The ShareesController doesn't have any unit or acceptance tests, so I'm unable to add them for this PR.

michielbdejong commented:
phpstan fixed by copying the pattern from https://github.com/owncloud/core/blob/v10.11.0/lib/private/Server.php#L896

michielbdejong commented:
I got my dev env into a state where I can run `bash tests/drone/test-phpunit.sh` - checking it now.

michielbdejong commented:
Can someone re-run "apiShareManagementToRoot-mariadb10.2-php7.4"?
It looks like it might be an intermittent failure: ["Step is killed (canceled)"](https://drone.owncloud.com/owncloud/core/37675/53/2)

phil-davis commented:
> Can someone re-run "apiShareManagementToRoot-mariadb10.2-php7.4"? It looks like it might be an intermittent failure: ["Step is killed (canceled)"](https://drone.owncloud.com/owncloud/core/37675/53/2)

The pipeline step hung for more than 4 hours. I guess drone eventually killed it. It looks like some drone agent problem happened.

I restarted the whole drone CI.

michielbdejong commented:
That helped, thanks a lot!

@phil-davis can you also do the review maybe, or should we try to find someone else available?

phil-davis commented:
@michielbdejong see #40587 - we can actually use a default call to `getSystemValue`. The problem was "just" that the mocking in the unit tests needs to know about config value `sharing.remoteShareesSearch`

Maybe you could cherry-pick the commit from there, and squash all the commits in this PR, and we can get it reviewed.

phil-davis commented on changelog/unreleased/40577:
```diff
@@ -0,0 +1,10 @@
+is installed, use it instead of the federatedfilesharing app to
+find sharee matches for OCM sharing.
+
+https://github.com/owncloud/core/issues/40576
```
```suggestion
https://github.com/owncloud/core/issues/40577
```

phil-davis reviewed (commented):
One tiny thing!

jvillafanez commented:
Technically, I'm ok with the solution. My worry is that neither core nor any ownCloud app uses that feature, so it's possible that we see it as dead code in the future. Any idea how we can make it more visible to prevent the feature from being removed? Maybe adding a comment saying that it's being used in the 3rd party sciencemesh app could be enough.

Another thing that worries me is that there is no documentation available about the expectation of the plugin class. If I (or anyone else) want to provide a custom implementation, I simply can't without seeing your code. I also don't know if your code works as expected. I mean, it could return some results, but I don't know if it should return more results or not.

It might be better if ownCloud core provides some kind of interface that the sciencemesh app can implement, and take advantage of the interface to document the expected behavior.
As far as I know, this should be only creating the interface somewhere in the "lib/public" folder (need to find a good place there), and the sciencemesh app just needs to implement the interface. You might want to adjust this PR a bit in order to ensure that the class being loaded implements the interface.

michielbdejong commented:
> neither core nor any ownCloud app uses that feature, so it's possible that we see it as dead code in the future

Ah no, but it's used by https://github.com/pondersource/oc-sciencemesh which will be installed on many production systems in various universities and research institutes across Europe.

DeepDiver1975 commented on apps/files_sharing/lib/Controller/ShareesController.php:
```diff
@@ -381,6 +381,15 @@ protected function getGroups($search) {
+		$pluginClass = $this->config->getSystemValue('sharing.remoteShareesSearch');
+		if ($pluginClass !== '') {
+			$this->result['remotes'] = [];
+			$plugin = new $pluginClass($this->config, $this->userManager, $this->userSession);
```
Ctor args might change over time. 
Can we use dependency injection here?

michielbdejong commented:
> It might be better if ownCloud core provides some kind of interface that the sciencemesh app can implement, and take advantage of the interface to document the expected behavior.

Good idea! I can add that.

jvillafanez commented:
>>  It might be better if ownCloud core provides some kind of interface that the sciencemesh app can implement, and take advantage of the interface to document the expected behavior.
>
> Good idea! I can add that.

In addition, make sure there is some kind of reference that the interface is being used and implemented by 3rd party apps so we know that we need to be careful with it. We might want to change the interface at some point and that would break your app, so a warning would help to prevent that problem in the future.

labkode commented:
@phil-davis the CI looks like it failed, can you please reload?

DeepDiver1975 commented on lib/public/Share/IRemoteShareesSearch.php:
```diff
@@ -0,0 +1,47 @@
+ * of this interface.
+ *
+ * @package OCP\Share
+ * @since 10.11.0
```
10.12.0 it will be

jvillafanez commented on apps/files_sharing/lib/Controller/ShareesController.php:
```diff
@@ -381,6 +382,18 @@ protected function getGroups($search) {
+		$pluginClass = $this->config->getSystemValue('sharing.remoteShareesSearch');
+		if ($pluginClass !== '') {
+			$this->result['remotes'] = [];
+			$app = new Application();
```
I don't know if this works, but it seems weird. I mean, the application instance is from the files_sharing app, so I expect it to be able to load classes from the files_sharing app, but not from any other app.

michielbdejong replied:
Ah, I thought there was only one DI container, not one per app?
But I would love to learn a better way to do this; how else would you suggest getting our hands on a `$pluginClass` instance?

I thought we cannot change the [`use OCA\Files_Sharing\AppInfo\Application;`](https://github.com/pondersource/core/blob/surf-dev/apps/files_sharing/lib/Controller/ShareesController.php#L44) statement to `use OCA\FederatedGroups\Application;` because that app might not be installed.

jvillafanez replied:
There should be a registration mechanism where your app could register an instance of your class. Your app is responsible of creating the object and then register the object so core or other apps can use it through a known interface.

There are 2 approaches for the registration mechanism:
* Core provides a class accessible to everyone so they can register the implementation (usually a "manager" class). The files_sharing app would get the implementation from this class.
* Either core or an app (files_sharing in this case) can throw an event and get the instances from there. Your app would need to be listening to that event and create the implementation at that moment. The files_sharing app would get the implementation from the event.

The first approach might be easier, although it could be questionable that core provides a manager that it won't use (it will be used by the files_sharing app). The additional drawback is that the manager will need to be instantiated when your app is loaded in order for you to register your class, but it shouldn't be too taxing.

The second approach will delay the instantiation until the last moment. The problem here is that the interface (your `IRemoteShareesSearch`) needs to be known by all the parties involved. Leaving it in core seems the easiest approach.

In any case, we need to take care of how the files_sharing app will get the right instance from the manager. There could be no instance registered, or there could be multiple instances from multiple apps.

michielbdejong replied:
Hm, I'm not sure whether I see the benefit of the change you are requesting.

We first implemented the functionality [in the same way as `sharing.managerFactory`](https://github.com/owncloud/core/blob/master/lib/private/Server.php#L898), directly using the `new` keyword with the classname from the config.

This worked, and is in line with existing core code of `lib/private/Server.php`.

But then last week you [asked us to change it](https://github.com/owncloud/core/pull/40577#discussion_r1081245090) to use DI instead. I agree this is an improvement, and we did what you asked us. But it was my understanding that this change would lead to the PR getting merged?

Are you saying that, on second thought, you think DI was not the way to go, and we should revert this change?

If we want to move away from DI then we can also change it back to the way it was when we first opened this PR, and maybe  to reduce the number of constructor arguments just pass an instance of `OC\Server`. That approach works well in [lib/private/Server.php](https://github.com/owncloud/core/blob/master/lib/private/Server.php#L898) so we can reuse it here, right?

jvillafanez replied:
Maybe I'm too used to my own way and I didn't explain it clear enough...

For the first approach, core would provide a class like:
```
class RemoteShareeSearcherManager implements IRemoteShareeSearcherManager {
  public function registerRemoteSharee(IRemoteShareeSearch $search) {
    // not sure if using the classname is a good approach or not
    $this->registered[\get_class($search)] = $search;
  }
  public function getRegisteredRemoteSharee($name) {......}
}
```
The ShareesController would still need to have this class injected in order to use it and get the configured searcher.
From your app, you'd get an instance of this RemoteShareeSearcherManager class and register a fresh instance of your RemoteShareesSearch.

------

For the second approach, the ShareesController would need to have an eventDispatcher injected (there should be examples in the code), and use that eventDispatcher to send an event and retrieve a fresh instance of the RemoteShareeSearch from the event.

Your app would need to listen to this event and create a fresh instance once the event is received. The created instance would be sent through the event.

------

The main point is that your app is responsible of creating your instance of the RemoteShareeSearch, so you can create it the way you want. Of course, you should use DI for this.
You still need to inject either the RemoteShareeSearcherManager or the eventDispatcher (depending on the chosen approach) into the ShareesController.

In both cases, no `new` keyword should be used outside of the DI. Maybe for the second approach we might need to be a bit more lenient and keep the usage of the `new` keyword within the Application instance.

The goal is to provide as much isolation as possible. This means that an app shouldn't depend on other apps and interact with core through interfaces. I know this isn't true in a lot of cases, but at least we should try.
Right now, the files_sharing app needs to create an instance of you RemoteShareesSearch implementation, even if it's done via dependency injection. This means an implicit dependency from the files_sharing app to your app, and it's something we should avoid as much as possible.

DeepDiver1975 replied:
We have multiple concepts which evolved over the past 10+ years ....

To simplify this a bit and also get this going:
```php
/** @var \OCP\Share\IRemoteShareesSearch $plugin */
$plugin = \OC::$server->query($pluginClass)

michielbdejong commented:
Rebased with signoff.
Added [documentation of optional server URL](https://github.com/owncloud/core/compare/8ea1badaf4bcbd39f3d24b86a00c2ab3cb599813..6c448ed4d5593395d74a5de411c7ebf7dd4d3f28)

phil-davis commented:
@jvillafanez I see that you are involved in some discussion above.
@DeepDiver1975 approved a few days ago.

Is this OK to merge now? Or?

jvillafanez commented:
I'll leave the final decision to @DeepDiver1975 . I haven't tested the solution, so I assume it works, but for me, that piece is "suspicious" in the sense that it seems weird to me and there are (in my opinion) technically better alternatives.
If everyone is ok with the current code, I'm not blocking it.

michielbdejong commented:
@DeepDiver1975 please let us know, I'm also fine with changing it a second time, as long as it doesn't become a moving target. :)

michielbdejong commented:
@DeepDiver1975 please let us know if there is anything else we can do, we are waiting for this to be merged asap if possible.

michielbdejong committed:
Get plugin from \OC::$server->query()

michielbdejong commented:
Great solution @DeepDiver1975, I didn't even know that was possible. So now the PR is super small and also it doesn't mention any particular app, awesome!
I tested it on GitPod in various way, and it all seems to work well. Thanks a lot!

phil-davis commented on changelog/unreleased/40577:
```diff
@@ -0,0 +1,10 @@
+Enhancement: Add support for OCM via ScienceMesh
+
+
+We've added an if-statement in the files_sharing ShareesController
+code that searches for remote sharees. When the 'sciencemesh' app
+is installed, use it instead of the federatedfilesharing app to
+find sharee matches for OCM sharing.
```
Should the description be adjusted to say something about `sharing.remoteShareesSearch` ?
Any app can define this in order to use this feature.

michielbdejong replied:
Done

michielbdejong committed:
Improve changelog entry

Signed-off-by: Michiel de Jong <michiel@unhosted.org>

michielbdejong committed:
improve changelog entry

Signed-off-by: Michiel de Jong <michiel@unhosted.org>

phil-davis commented:
code-style is complaining. It needs:
`make test-php-style-fix`
and then commit and push the changes.

I tried to do it, but I can't push to this branch in `pondersource` repo.

michielbdejong commented:
Ah yes, [b8f9fdf](https://github.com/owncloud/core/pull/40577/commits/b8f9fdfa02614a0832bb5992d38a0a9ef2d8d4be) broke it! Sorry.

michielbdejong commented:
In https://drone.owncloud.com/owncloud/core/37988/28/2 the 'skip-if-unchanged' step of Sonar failed, apparently unrelated to this PR. I'll try what happens if I push a whitespace change.

michielbdejong committed:
Whitespace change to re-trigger CI

Signed-off-by: Michiel de Jong <michiel@unhosted.org>

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiFederationToShares1-latest-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/37990/100
<!-- id: pr-40577 -->

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiFederationToShares1-git-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/37990/99
<!-- id: pr-40577 -->

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiSharingNotificationsToRoot-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/37990/91
<!-- id: pr-40577 -->

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiWebdavUpload2-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/37990/89
<!-- id: pr-40577 -->

ownclouders commented:
:boom: Acceptance tests pipeline <strong>cliBackground-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/37990/106
<!-- id: pr-40577 -->

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiWebdavUpload1-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/37990/88
<!-- id: pr-40577 -->

michielbdejong commented:
@phil-davis what is the next step now?

phil-davis commented:
https://drone.owncloud.com/owncloud/core/37990/88/6
`Error response from daemon: No such image: owncloudci/php:7.4`
Stupid drone+docker gets it wrong. I have restarted CI.

phil-davis reviewed (approved):
LGTM

phil-davis commented:
@DeepDiver1975 please review again. IMO this is ready to merge.

michielbdejong commented:
@DeepDiver1975 Please merge this today if possible.

phil-davis commented:
The CI fail is https://drone.owncloud.com/owncloud/core/38013/28/2
sonar-analysis
```
latest: Pulling from owncloudci/drone-skip-pipeline
Digest: sha256:78d133bf897c26ab34c17839d1ffe86b15174f12370745cbab8b949098620174
Status: Image is up to date for owncloudci/drone-skip-pipeline:latest
DRONE_COMMIT_BEFORE:  13d0a217480ceba2846037ad782624e6645bb391
time="2023-02-24T04:51:17Z" level=error msg="execution failed: commit from DRONE_COMMIT_BEFORE not found: object not found"
```
I suppose that will be something to do with the fork that this branch comes from and???

But all tests passed.

michielbdejong commented:
Yes, it's about putting the `{` on the next line. But when you fix Sonar then `make test-php-style-fix` breaks and vice versa, so then I reverted that.

michielbdejong commented:
@DeepDiver1975 ? Anyone else?

jnweiger commented:
https://drone.owncloud.com/owncloud/core/38013
explodes in sonar analysis with `level=error msg="execution failed: commit from DRONE_COMMIT_BEFORE not found: object not found"`
Never seen that before.

For inclusion in 10.12.0, please redirect this PR to the `release-10.12.0` branch (instead of `master`)

phil-davis commented:
> For inclusion in 10.12.0, please redirect this PR to the release-10.12.0 branch (instead of master)

See PR #40667

phil-davis commented:
Merged in #40667
